### PR TITLE
DecoSchematic: Check for non-walkable node above placement position

### PR DESCRIPTION
**Pull request by paramat**
See Issue #3640

### CODE REVIEW

**Comment by paramat** [comment]
Hmm not sure this is a good idea as it is only effective for tree schematics with central trunks. Trees placed on top of each other are rare and this code slows down schematic placement.

**Comment by est31** [comment]
Why are the trees placed on top of each other the first place?

**Comment by paramat** [comment]
Decoration placement uses the mapgen heightmap, but if the heightmap isn't present it uses 'findGroundLevel' or 'findLiquidSurface', these scan down a column to find the first walkable node, or the first liquid node, this ensures there is air or water above the placement position.

Since heightmaps are present for all mapgens, schematic decorations are placed at the surface node at the heightmap y, without checking if there is air above, i assume hmmmm had reason to do it this way, maybe a similar reason to my post above.

I'm still undecided this is worth fixing.

File: src/mg_decoration.cpp [hunk 1]
```diff
@@ -294,7 +294,7 @@ bool DecoSimple::canPlaceDecoration(MMVManip *vm, v3s16 p)
 }
 
 
-size_t DecoSimple::generate(MMVManip *vm, PcgRandom *pr, v3s16 p)
+size_t DecoSimple::generate(Mapgen *mg, MMVManip *vm, PcgRandom *pr, v3s16 p)
```
Comment (kwolekr) [inline src/mg_decoration.cpp]: Wait, why are you adding the whole mapgen to the Schematic interface?  This violates the principle of least access.

I think you should be able to access the member `ndef` or `m_ndef` anyway as it should be inherited from ObjDef.

File: src/mg_decoration.cpp [hunk 2]
```diff
@@ -347,6 +347,16 @@ size_t DecoSchematic::generate(MMVManip *vm, PcgRandom *pr, v3s16 p)
...
 		return 0;
 
+	// Check for a non-walkable node above, to be consistent with the
+	// behaviour of findGroundLevel and findLiquidSurface, which are
+	// used if the heightmap is not present. This avoids, for example,
+	// multiple tree trunks placed at the same position.
+	v3s16 em = vm->m_area.getExtent();
+	vm->m_area.add_y(em, vi, 1);
+	MapNode &n = vm->m_data[vi];
+	if (mg->ndef->get(n).walkable)
```
Comment (kwolekr) [inline src/mg_decoration.cpp]: This kind of hard coded logic leaves me scratching my head.  I'm not sure if checking the node above for walkable is really the right thing to do to begin with.

**Comment by kwolekr** [comment]
Hmm, unless I'm mistaken, it seems like the real problem is that the heightmap isn't updated in between decoration placements.  Is checking for "walkable" really the correct approach??

**Comment by paramat** [comment]
My reason for checking 'walkable' is explained in the issue.
I think i'll close this and re-add roots to mapgen trees to solve it in a simpler way.

**Comment by paramat** [comment]
See https://github.com/minetest/minetest_game/pull/825

**Pull request closed without merging**